### PR TITLE
Feature: add full bootstrap on-policy actor loss

### LINKED ISSUE #41: Build ops for computing MLE model learning loss

### LINKED ISSUE #42: Build ops for optimizing the dynamics model parameters via MLE

### PR DESCRIPTION
**Pull request by 0xangelo**
This PR includes:
* Full bootstrap, model-aware, on-policy actor loss
* Model learning in on-policy mapo
* Tests of MAPOModel

Closes #41 
Closes #42

### CODE REVIEW

File: mapo/agents/mapo/mapo_policy.py [hunk 1]
```diff
@@ -10,23 +12,67 @@
 from ray.rllib.evaluation.postprocessing import compute_advantages, Postprocessing
 
 
-def build_actor_critic_losses(policy, batch_tensors):
-    """Contruct actor (DPG) and critic (Fitted Q) tf losses."""
+AgentComponents = namedtuple("AgentComponents", "dynamics critic actor")
```
Comment (thiagopbueno) [inline mapo/agents/mapo/mapo_policy.py:15]: What about `AgentModel` ?

File: mapo/agents/mapo/tests/test_model.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+def get_models():
+    obs_space, action_space = get_spaces()
+    model_config = {
+        "custom_options": {
+            "actor": {"activation": "relu", "layers": [32, 32]},
+            "critic": {"activation": "relu", "layers": [32, 32]},
+            "dynamics": {"activation": "relu", "layers": [32, 32]},
+        }
+    }
+    models = [
```
Comment (thiagopbueno) [inline mapo/agents/mapo/tests/test_model.py:29]: Just a suggestion... you may find convenient to use `itertools.product([True, False], [True, False])` to generate all cross product combinations for the `target_networks` and `twin_q` aguments... something along the lines of:

```python3
models = [
    MAPOModel(
            obs_space,
            action_space,
            1,
            model_config,
            "test_model",
            target_networks=target_networks,
            twin_q=twin_q
    )
    for target_networks, twin_q in itertools.product([False, True], [False, True])
]
```

File: mapo/models/dynamics/tests/test_gaussian.py [hunk 3]
```diff
@@ -167,6 +167,14 @@ def test_gaussian_sample(model, state, action):
     assert next_state.dtype == state.dtype
 
 
+@pytest.mark.parametrize("model", get_models())
+def test_gaussian_sample_propagates_gradients(model, state, action):
```
Comment (thiagopbueno) [inline mapo/models/dynamics/tests/test_gaussian.py:171]: So, this is the usual behavior in `tf.distributions`... if the distribution is re-parametrizable then you can indeed backpropagate through the sampling procedure...

It would be nice to add the complementary test here.. to check that once we gradient-stop the sample, trying to get back the gradient will return None (or an error...).

File: mapo/agents/mapo/mapo_policy.py [hunk 4]
```diff
@@ -10,23 +12,67 @@
...
+def _build_actor_loss(batch_tensors, model, config):
+    obs = batch_tensors[SampleBatch.CUR_OBS]
+    gamma = config["gamma"]
+    n_samples = config["branching_factor"]
+    policy_action = model.compute_actions(obs)
+    sampled_next_state, next_state_log_prob = model.compute_log_prob_sampled(
+        obs, policy_action, (n_samples,)
+    )
+    next_state_value = tf.stop_gradient(model.compute_state_values(sampled_next_state))
+    model_aware_policy_loss = tf.reduce_mean(
```
Comment (thiagopbueno) [inline mapo/agents/mapo/mapo_policy.py:48]: BTW.. I added a `_reward_fn` (and also a `_transition_fn`) method in NavigationEnv that takes state and action tensors and return another tensor for the reward computation... I will create a base class to be used for all custom envs that properly implements this API.. but by now we should be able to compute the `mapo_actor_loss` with the reward gradient term... 

Just need to figure it out a way to get access to the environment in the policy... any thoughts?

**Review by thiagopbueno — CHANGES_REQUESTED** [review]

File: mapo/agents/mapo/tests/test_model.py (hunk 2 above)
Comment (0xangelo) [reply mapo/agents/mapo/tests/test_model.py:29]: Sounds good!

File: mapo/models/dynamics/tests/test_gaussian.py (hunk 3 above)
Comment (0xangelo) [reply mapo/models/dynamics/tests/test_gaussian.py:171]: Yeah, I'll add it as a "sanity check".

File: mapo/agents/mapo/mapo_policy.py (hunk 1 above)
Comment (0xangelo) [reply mapo/agents/mapo/mapo_policy.py:15]: I just wanted something to name tuple components by the models they appertain to. I couldn't think of a better name.

File: mapo/agents/mapo/mapo_policy.py (hunk 4 above)
Comment (0xangelo) [reply mapo/agents/mapo/mapo_policy.py:48]: For now, I can think of two ways:
* Add a callback on episode start that calls a custom method (maybe `set_current_env`) to set the environment the policy is using. Callbacks are passed through the config (`config["callbacks"]["on_episode_start"]`), but we can add then in `MAPOTrainer` so that they're added every time ([DQN](https://github.com/ray-project/ray/blob/e4854b11f9a032261a3a07029b8492e59002da61/python/ray/rllib/agents/dqn/dqn.py#L168) does this). The callback is called with the following arguments:
```python
            callbacks["on_episode_start"]({
                "env": base_env,
                "policy": policies,
                "episode": episode,
            })
```
* Create a separate environment in the policy using `config["env"]` and `config["env_config"]`. This can be problematic if the environment has some hidden internal state, since in that case the instance used for calculation transitions and the one used for training might behave differently. Nevertheless, we would probably do something similar to what `Trainer` does:
```python
    @override(Trainable)
    def _setup(self, config):
        env = self._env_id
        if env:
            config["env"] = env
            if _global_registry.contains(ENV_CREATOR, env):
                self.env_creator = _global_registry.get(ENV_CREATOR, env)
            else:
                import gym  # soft dependency
                self.env_creator = lambda env_config: gym.make(env)
        else:
            self.env_creator = lambda env_config: None

```
